src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.cs

```diff
@@ -326,20 +326,232 @@ private void MakeAndAssignEnvironments()
                     var env = new ClosureEnvironment(variablesInEnvironment, isStruct);
                     scope.DeclaredEnvironments.Add(env);
 
                     _topLevelMethod.TryGetThisParameter(out var thisParam);
+                    var capturingClosures = new HashSet<Closure>();
                     foreach (var closure in closures)
                     {
+                        capturingClosures.Add(closure);
+
                         closure.CapturedEnvironments.Add(env);
                         if (thisParam != null && env.CapturedVariables.Contains(thisParam))
                         {
                             closure.CapturesThis = true;
                         }
                     }
+
+                    closuresCapturingScopeVariables.Add(scope, capturingClosures);
                 });
+
+                _closuresCapturingScopeVariables = closuresCapturingScopeVariables;
+            }
+
+            /// <summary>
+            /// Must be called only after <see cref="MakeAndAssignEnvironments"/> and <see cref="ComputeLambdaScopesAndFrameCaptures"/>
+            /// 
+            /// In order to reduce allocations, merge environments into a parent environment when it is safe to do so.
+            /// This must be done whilst preserving semantics.
+            /// 
+            /// We also have to make sure not to extend the life of any variable.
+            /// This means that we can only merge an environment into its parent if exactly the same closures directly or indirectly reference both environments.
+            /// </summary>
+            private void MergeEnvironments()
+            {
+                // for now we don't analyze jumps to check if it is safe to optimize, and instead return.
+                if (containsJumps(ScopeTree.BoundNode.Syntax))
+                    return;
+
+                // if a closure captures a scope, which captures its parent, then the closure also captures the parents scope.
+                // we update _closuresCapturingScopeVariables to reflect this.
+                foreach (var (scope, capturingClosures) in _closuresCapturingScopeVariables)
+                {
+                    if (scope.DeclaredEnvironments.Count == 0)
+                        continue;
+                    // Right now we only create one environment per scope
+                    Debug.Assert(scope.DeclaredEnvironments.Count == 1);
+                    var env = scope.DeclaredEnvironments[0];
+
+                    if (env.CapturesParent)
+                    {
+                        var currentScope = scope.Parent;
+                        while (true)
+                        {
+                            if (currentScope == null)
+                            {
+                                throw ExceptionUtilities.Unreachable;
+                            }
+
+                            if (currentScope.DeclaredEnvironments.Count == 0 || currentScope.DeclaredEnvironments[0].IsStruct)
```

Comment:
IIRC capturing a struct environment should be illegal. Isn't this represented by a closure capturing each struct environment individually, instead of one environment capturing another (like in the case of class environments)?

Reply:
So I should verify that this is the case and either or remove the check, or assert that the it is not a struct?

Reply:
As far as I can make out from a quick check, whilst we can't capture a struct environment, we can capture past a struct environment, so there maybe a class environment which captures a class environment, and there is a struct environment in between them.

That might be impossible in practice, but ComputeLambdaScopesAndFrameCaptures seems to allow for it.

Reply:
Good point, that is definitely possible.